### PR TITLE
Do not add NULLS FIRST/LAST for NOT NULL columns (pg)

### PR DESCRIPTION
**Pull request by lzap**
It looks like postgresql optimizer is not able to deal with situation when a
column is NOT NULL but NULLS FIRST/LAST is provided. In that case, index is not
used at all which is not efficient.

If you like my patch, I can try to add tests, but I struggle with running them
locally. Hopefully travis will help me out.

### CODE REVIEW

**Comment by wvanbergen** [comment]
Feels like a bug in the PostgreSQL query planner, so I suppose we'll have to work around it.

The approach looks good to me, but apparently the tests are failing with `the field '' in the order statement is not valid field for search`. That probably means you are returning an empty string instead of a `nil` value somewhere.

**Comment by lzap** [comment]
Yeah thanks for info. Since I am unable to setup my dev environment and debug this, I am doing trial and error via travis. Do you want me to write a test? Not quite sure how it should look like to be honest :-)

**Comment by wvanbergen** [comment]
If you have Postgres running locally (or elsewhere), it shouldn't be too hard. Edit https://github.com/wvanbergen/scoped_search/blob/master/spec/database.ruby.yml to match your credentials / databases. You can comment out the sqlite and mysql sections so the tests won't run against those DBs.

File: lib/scoped_search/query_builder.rb [hunk 1]
```diff
@@ -83,16 +83,20 @@ def build_find_params(options)
       return find_attributes
     end
 
-    def order_by(order, &block)
+    def find_field_for_order_by(order, &block)
       order ||= definition.default_order
       return nil if order.blank?
```
Comment (domcleal) [inline lib/scoped_search/query_builder.rb]: This used to short-circuit exit the order_by method but no longer does as it's in a different method now.  Maybe add a check in `order_by` after calling `find_field_for_order_by` to `return nil if field.nil?`

File: lib/scoped_search/query_builder.rb [hunk 2]
```diff
@@ -516,7 +520,10 @@ def to_not_sql(rhs, definition, &block)
 
       def order_by(order, &block)
         sql = super(order, &block)
-        sql += sql.include?('DESC') ? ' NULLS LAST ' : ' NULLS FIRST ' if sql
+        if sql
+          field, _ = find_field_for_order_by(order, &block)
+          sql += sql.include?('DESC') ? ' NULLS LAST ' : ' NULLS FIRST ' if field.column.null
```
Comment (domcleal) [inline lib/scoped_search/query_builder.rb]: Probably handle `field` being nil here too.

**Comment by lzap** [comment]
Wow tests are green! Thanks @domcleal!

Do I understand the test codebase correctly that this is actually covered by pgsql integration tests? No need of unit test then? :-)

**Comment by domcleal** [comment]
I think it's exercised against PostgreSQL, but I'm not sure that the NULLS LAST/FIRST behaviour is actually checked anywhere (i.e. regression check for value ordering when the column accepts NULL).  Otherwise it reads well to me.

**Comment by wvanbergen** [comment]
I think the `NULLS FIRST/LAST` part was explicitly added to make test results pass for both MySQL and PostgreSQL. So IIRC this is covered by the integration tests

**Comment by wvanbergen** [comment]
Thanks for fixing this!

**Comment by domcleal** [comment]
Thanks @lzap!

@wvanbergen do you plan on a patch release soon?  There are a few useful bug fixes in master.

**Comment by wvanbergen** [comment]
will do

**Comment by domcleal** [comment]
Cheers.

**Pull request merged**